File: android/src/main/java/com/netease/nertcflutter/NERtcStatsObserverImpl.java
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

package com.netease.nertcflutter;

import com.netease.lava.nertc.sdk.audio.NERtcAudioStreamType;
import com.netease.lava.nertc.sdk.stats.NERtcAudioLayerRecvStats;
import com.netease.lava.nertc.sdk.stats.NERtcAudioLayerSendStats;
import com.netease.lava.nertc.sdk.stats.NERtcAudioRecvStats;
import com.netease.lava.nertc.sdk.stats.NERtcAudioSendStats;
import com.netease.lava.nertc.sdk.stats.NERtcNetworkQualityInfo;
import com.netease.lava.nertc.sdk.stats.NERtcStats;
import com.netease.lava.nertc.sdk.stats.NERtcStatsObserver;
import com.netease.lava.nertc.sdk.stats.NERtcVideoLayerRecvStats;
import com.netease.lava.nertc.sdk.stats.NERtcVideoLayerSendStats;
import com.netease.lava.nertc.sdk.stats.NERtcVideoRecvStats;
import com.netease.lava.nertc.sdk.stats.NERtcVideoSendStats;
import io.flutter.plugin.common.BinaryMessenger;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class NERtcStatsObserverImpl implements NERtcStatsObserver {

  private final Messages.NERtcStatsEventSink _eventSink;
  private final String channelTag;

  NERtcStatsObserverImpl(BinaryMessenger argBinaryMessenger, String channelTag) {
    _eventSink = new Messages.NERtcStatsEventSink(argBinaryMessenger);
    this.channelTag = channelTag;
  }

  @Override
  public void onRtcStats(NERtcStats stat) {
    if (stat == null) return;
    HashMap<Object, Object> stats = toMap(stat);
    _eventSink.onRtcStats(
        stats,
        channelTag,
        aVoid -> {
          // 实现reply方法
        });
  }

  @Override
  public void onLocalAudioStats(NERtcAudioSendStats stat) {
    if (stat == null) return;
    Map<Object, Object> statMap = toMap(stat);
    if (statMap != null) {
      _eventSink.onLocalAudioStats(
          statMap,
          channelTag,
          aVoid -> {
            // 实现reply方法
          });
    }
  }

  @Override
  public void onRemoteAudioStats(NERtcAudioRecvStats[] stats) {
    if (stats != null && stats.length > 0) {
      List<Map<Object, Object>> statList = new ArrayList<>(stats.length);
      for (NERtcAudioRecvStats stat : stats) {
        Map<Object, Object> statMap = toMap(stat);
        if (statMap != null) {
          statList.add(statMap);
        }
      }
      HashMap<Object, Object> map = new HashMap<>();
      map.put("list", statList);
      _eventSink.onRemoteAudioStats(
          map,
          channelTag,
          aVoid -> {
            // 实现reply方法
          });
    }
  }

  @Override
  public void onLocalVideoStats(NERtcVideoSendStats stat) {
    if (stat == null) return;
    HashMap<Object, Object> map = toMap(stat);
    _eventSink.onLocalVideoStats(
        map,
        channelTag,
        aVoid -> {
          // 实现reply方法
        });
  }

  @Override
  public void onRemoteVideoStats(NERtcVideoRecvStats[] stats) {
    if (stats != null && stats.length > 0) {
      List<Map<String, Object>> statList = new ArrayList<>(stats.length);
      for (NERtcVideoRecvStats stat : stats) {
        statList.add(toMap(stat));
      }
      HashMap<Object, Object> map = new HashMap<>();
      map.put("list", statList);
      _eventSink.onRemoteVideoStats(
          map,
          channelTag,
          aVoid -> {
            // 实现reply方法
          });
    }
  }

  @Override
  public void onNetworkQuality(NERtcNetworkQualityInfo[] stats) {
    if (stats != null && stats.length > 0) {
      List<Map<String, Object>> statList = new ArrayList<>(stats.length);
      for (NERtcNetworkQualityInfo stat : stats) {
        statList.add(toMap(stat));
      }
      HashMap<Object, Object> map = new HashMap<>();
      map.put("list", statList);
      _eventSink.onNetworkQuality(
          map,
          channelTag,
          aVoid -> {
            // 实现reply方法
          });
    }
  }

  private HashMap<String, Object> toMap(NERtcNetworkQualityInfo info) {
    HashMap<String, Object> map = new HashMap<>();
    map.put("uid", info.userId);
    map.put("txQuality", info.upStatus);
    map.put("rxQuality", info.downStatus);
    return map;
  }

  private HashMap<Object, Object> toMap(NERtcStats stats) {
    HashMap<Object, Object> map = new HashMap<>();
    map.put("txBytes", stats.txBytes);
    map.put("rxBytes", stats.rxBytes);
    map.put("cpuAppUsage", stats.cpuAppUsage);
    map.put("cpuTotalUsage", stats.cpuTotalUsage);
    map.put("memoryAppUsageRatio", stats.memoryAppUsageRatio);
    map.put("memoryTotalUsageRatio", stats.memoryTotalUsageRatio);
    map.put("memoryAppUsageInKBytes", stats.memoryAppUsageInKBytes);
    map.put("totalDuration", stats.totalDuration);
    map.put("txAudioBytes", stats.txAudioBytes);
    map.put("txVideoBytes", stats.txVideoBytes);
    map.put("rxAudioBytes", stats.rxAudioBytes);
    map.put("rxVideoBytes", stats.rxVideoBytes);
    map.put("rxAudioKBitRate", stats.rxAudioKBitRate);
    map.put("rxVideoKBitRate", stats.rxVideoKBitRate);
    map.put("txAudioKBitRate", stats.txAudioKBitRate);
    map.put("txVideoKBitRate", stats.txVideoKBitRate);
    map.put("upRtt", stats.upRtt);
    map.put("downRtt", stats.downRtt);
    map.put("txAudioPacketLossRate", stats.txAudioPacketLossRate);
    map.put("txVideoPacketLossRate", stats.txVideoPacketLossRate);
    map.put("txAudioPacketLossSum", stats.txAudioPacketLossSum);
    map.put("txVideoPacketLossSum", stats.txVideoPacketLossSum);
    map.put("txAudioJitter", stats.txAudioJitter);
    map.put("txVideoJitter", stats.txVideoJitter);
    map.put("rxAudioPacketLossRate", stats.rxAudioPacketLossRate);
    map.put("rxVideoPacketLossRate", stats.rxVideoPacketLossRate);
    map.put("rxAudioPacketLossSum", stats.rxAudioPacketLossSum);
    map.put("rxVideoPacketLossSum", stats.rxVideoPacketLossSum);
    map.put("rxAudioJitter", stats.rxAudioJitter);
    map.put("rxVideoJitter", stats.rxVideoJitter);
    return map;
  }

  private HashMap<Object, Object> toMap(NERtcAudioSendStats stats) {
    HashMap<Object, Object> statsMap = new HashMap<>();
    ArrayList<HashMap<Object, Object>> layers = new ArrayList<>();
    for (NERtcAudioLayerSendStats layer : stats.audioLayers) {
      HashMap<Object, Object> map = new HashMap<>();
      map.put("streamType", layer.streamType.ordinal());
      map.put("kbps", layer.kbps);
      map.put("lossRate", layer.lossRate);
      map.put("rtt", layer.rtt);
      map.put("volume", layer.volume);
      map.put("numChannels", layer.numChannels);
      map.put("sentSampleRate", layer.sentSampleRate);
      map.put("capVolume", layer.capVolume);
      layers.add(map);
    }
    statsMap.put("layers", layers);
    return statsMap;
  }

  private HashMap<Object, Object> toMap(NERtcAudioRecvStats stats) {
    HashMap<Object, Object> map = new HashMap<>();
    map.put("uid", stats.uid);
    ArrayList<HashMap<String, Object>> layers = new ArrayList<>();
    if (stats.layers != null) {
      for (NERtcAudioLayerRecvStats stat : stats.layers) {
        HashMap<String, Object> mapLayer = new HashMap<>();
        int type = (stat.streamType == NERtcAudioStreamType.kNERtcAudioStreamTypeMain) ? 0 : 1;
        mapLayer.put("kbps", stat.kbps);
        mapLayer.put("lossRate", stat.lossRate);
        mapLayer.put("volume", stat.volume);
        mapLayer.put("totalFrozenTime", stat.totalFrozenTime);
        mapLayer.put("frozenRate", stat.frozenRate);
        mapLayer.put("streamType", type);
        layers.add(mapLayer);
      }
    }
    map.put("list", layers);
    return map;
  }

  private HashMap<Object, Object> toMap(NERtcVideoSendStats stats) {
    HashMap<Object, Object> map = new HashMap<>();
    ArrayList<HashMap<String, Object>> layers = new ArrayList<>();
    if (stats.videoLayers != null) {
      for (NERtcVideoLayerSendStats stat : stats.videoLayers) {
        HashMap<String, Object> mapLayer = new HashMap<>();
        mapLayer.put("layerType", stat.layerType);
        mapLayer.put("captureWidth", stat.capWidth);
        mapLayer.put("captureHeight", stat.capHeight);
        mapLayer.put("width", stat.width);
        mapLayer.put("height", stat.height);
        mapLayer.put("sendBitrate", stat.sendBitrate);
        mapLayer.put("encoderOutputFrameRate", stat.encoderOutputFrameRate);
        mapLayer.put("captureFrameRate", stat.captureFrameRate);
        mapLayer.put("targetBitrate", stat.targetBitrate);
        mapLayer.put("encoderBitrate", stat.encoderBitrate);
        mapLayer.put("sentFrameRate", stat.sentFrameRate);
        mapLayer.put("renderFrameRate", stat.renderFrameRate);
        mapLayer.put("encoderName", stat.encoderName);
        mapLayer.put("dropBwStrategyEnabled", stat.dropBwStrategyEnabled);
        layers.add(mapLayer);
      }
    }
    map.put("layers", layers);
    return map;
  }

  private HashMap<String, Object> toMap(NERtcVideoRecvStats stats) {
    HashMap<String, Object> map = new HashMap<>();
    map.put("uid", stats.uid);
    ArrayList<HashMap<String, Object>> layers = new ArrayList<>();
    if (stats.layers != null) {
      for (NERtcVideoLayerRecvStats stat : stats.layers) {
        HashMap<String, Object> mapLayer = new HashMap<>();
        mapLayer.put("layerType", stat.layerType);
        mapLayer.put("width", stat.width);
        mapLayer.put("height", stat.height);
        mapLayer.put("receivedBitrate", stat.receivedBitrate);
        mapLayer.put("fps", stat.fps);
        mapLayer.put("packetLossRate", stat.packetLossRate);
        mapLayer.put("decoderOutputFrameRate", stat.decoderOutputFrameRate);
        mapLayer.put("rendererOutputFrameRate", stat.rendererOutputFrameRate);
        mapLayer.put("totalFrozenTime", stat.totalFrozenTime);
        mapLayer.put("frozenRate", stat.frozenRate);
        mapLayer.put("decoderName", stat.decoderName);
        layers.add(mapLayer);
      }
    }
    map.put("layers", layers);
    return map;
  }
}

File: ios/Classes/NERtcStatsObserverImpl.m
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

#import "NERtcStatsObserverImpl.h"

@interface NERtcStatsObserverImpl ()

@property(nonatomic, strong) NEFLTNERtcStatsEventSink *eventSink;
@property(nonatomic, copy) NSString *channelTag;

@end

@implementation NERtcStatsObserverImpl

- (instancetype)initWithBinaryMessenger:(id<FlutterBinaryMessenger>)binaryMessenger
                             channelTag:(NSString *)channelTag {
  self = [super init];
  if (self) {
    _eventSink = [[NEFLTNERtcStatsEventSink alloc] initWithBinaryMessenger:binaryMessenger];
    _channelTag = [channelTag copy];
  }
  return self;
}

#pragma mark - NERtcEngineMediaStatsObserver

- (void)onRtcStats:(NERtcStats *)stat {
#ifdef DEBUG
  NSLog(@"NERtcStatsObserverImpl:onRtcStats");
#endif
  if (!stat) return;

  NSDictionary *statsDic = [self toRtcStats:stat];
  [_eventSink onRtcStatsArguments:statsDic
                       channelTag:_channelTag
                       completion:^(FlutterError *_Nullable error) {
                         if (error) {
                           NSLog(@"onRtcStats error: %@", error.message);
                         }
                       }];
}

- (void)onLocalAudioStat:(NERtcAudioSendStats *)stat {
#ifdef DEBUG
  NSLog(@"NERtcStatsObserverImpl:onLocalAudioStat");
#endif
  if (!stat) return;

  NSDictionary *statsDic = [self toLocalAudioStats:stat];
  [_eventSink onLocalAudioStatsArguments:statsDic
                              channelTag:_channelTag
                              completion:^(FlutterError *_Nullable error) {
                                if (error) {
                                  NSLog(@"onLocalAudioStat error: %@", error.message);
                                }
                              }];
}

- (void)onRemoteAudioStats:(NSArray<NERtcAudioRecvStats *> *)stats {
#ifdef DEBUG
  NSLog(@"NERtcStatsObserverImpl:onRemoteAudioStats");
#endif
  if (!stats || stats.count == 0) return;

  NSMutableArray<NSDictionary *> *statsList = [[NSMutableArray alloc] init];
  for (NERtcAudioRecvStats *stat in stats) {
    NSDictionary *statDic = [self toRemoteAudioStats:stat];
    if (statDic) {
      [statsList addObject:statDic];
    }
  }

  NSDictionary *resultDic = @{@"list" : statsList};
  [_eventSink onRemoteAudioStatsArguments:resultDic
                               channelTag:_channelTag
                               completion:^(FlutterError *_Nullable error) {
                                 if (error) {
                                   NSLog(@"onRemoteAudioStats error: %@", error.message);
                                 }
                               }];
}

- (void)onLocalVideoStat:(NERtcVideoSendStats *)stat {
#ifdef DEBUG
  NSLog(@"NERtcStatsObserverImpl:onLocalVideoStat");
#endif
  if (!stat) return;

  NSDictionary *statsDic = [self toLocalVideoStats:stat];
  [_eventSink onLocalVideoStatsArguments:statsDic
                              channelTag:_channelTag
                              completion:^(FlutterError *_Nullable error) {
                                if (error) {
                                  NSLog(@"onLocalVideoStat error: %@", error.message);
                                }
                              }];
}

- (void)onRemoteVideoStats:(NSArray<NERtcVideoRecvStats *> *)stats {
#ifdef DEBUG
  NSLog(@"NERtcStatsObserverImpl:onRemoteVideoStats");
#endif
  if (!stats || stats.count == 0) return;

  NSMutableArray<NSDictionary *> *statsList = [[NSMutableArray alloc] init];
  for (NERtcVideoRecvStats *stat in stats) {
    NSDictionary *statDic = [self toRemoteVideoStats:stat];
    if (statDic) {
      [statsList addObject:statDic];
    }
  }

  NSDictionary *resultDic = @{@"list" : statsList};
  [_eventSink onRemoteVideoStatsArguments:resultDic
                               channelTag:_channelTag
                               completion:^(FlutterError *_Nullable error) {
                                 if (error) {
                                   NSLog(@"onRemoteVideoStats error: %@", error.message);
                                 }
                               }];
}

- (void)onNetworkQuality:(NSArray<NERtcNetworkQualityStats *> *)stats {
#ifdef DEBUG
  NSLog(@"NERtcStatsObserverImpl:onNetworkQuality");
#endif
  if (!stats || stats.count == 0) return;

  NSMutableArray<NSDictionary *> *statsList = [[NSMutableArray alloc] init];
  for (NERtcNetworkQualityStats *stat in stats) {
    NSDictionary *statDic = [self toNetworkQualityStats:stat];
    if (statDic) {
      [statsList addObject:statDic];
    }
  }

  NSDictionary *resultDic = @{@"list" : statsList};
  [_eventSink onNetworkQualityArguments:resultDic
                             channelTag:_channelTag
                             completion:^(FlutterError *_Nullable error) {
                               if (error) {
                                 NSLog(@"onNetworkQuality error: %@", error.message);
                               }
                             }];
}

#pragma mark - Private Methods

- (NSDictionary *)toRtcStats:(NERtcStats *)stat {
  NSMutableDictionary *map = [[NSMutableDictionary alloc] init];
  [map setObject:@(stat.txBytes) forKey:@"txBytes"];
  [map setObject:@(stat.rxBytes) forKey:@"rxBytes"];
  [map setObject:@(stat.cpuAppUsage) forKey:@"cpuAppUsage"];
  [map setObject:@(stat.cpuTotalUsage) forKey:@"cpuTotalUsage"];
  [map setObject:@(stat.memoryAppUsageRatio) forKey:@"memoryAppUsageRatio"];
  [map setObject:@(stat.memoryTotalUsageRatio) forKey:@"memoryTotalUsageRatio"];
  [map setObject:@(stat.memoryAppUsageInKBytes) forKey:@"memoryAppUsageInKBytes"];
  [map setObject:@(stat.totalDuration) forKey:@"totalDuration"];
  [map setObject:@(stat.txAudioBytes) forKey:@"txAudioBytes"];
  [map setObject:@(stat.txVideoBytes) forKey:@"txVideoBytes"];
  [map setObject:@(stat.rxAudioBytes) forKey:@"rxAudioBytes"];
  [map setObject:@(stat.rxVideoBytes) forKey:@"rxVideoBytes"];
  [map setObject:@(stat.rxAudioKBitRate) forKey:@"rxAudioKBitRate"];
  [map setObject:@(stat.rxVideoKBitRate) forKey:@"rxVideoKBitRate"];
  [map setObject:@(stat.txAudioKBitRate) forKey:@"txAudioKBitRate"];
  [map setObject:@(stat.txVideoKBitRate) forKey:@"txVideoKBitRate"];
  [map setObject:@(stat.upRtt) forKey:@"upRtt"];
  [map setObject:@(stat.downRtt) forKey:@"downRtt"];
  [map setObject:@(stat.txAudioPacketLossRate) forKey:@"txAudioPacketLossRate"];
  [map setObject:@(stat.txVideoPacketLossRate) forKey:@"txVideoPacketLossRate"];
  [map setObject:@(stat.txAudioPacketLossSum) forKey:@"txAudioPacketLossSum"];
  [map setObject:@(stat.txVideoPacketLossSum) forKey:@"txVideoPacketLossSum"];
  [map setObject:@(stat.txAudioJitter) forKey:@"txAudioJitter"];
  [map setObject:@(stat.txVideoJitter) forKey:@"txVideoJitter"];
  [map setObject:@(stat.rxAudioPacketLossRate) forKey:@"rxAudioPacketLossRate"];
  [map setObject:@(stat.rxVideoPacketLossRate) forKey:@"rxVideoPacketLossRate"];
  [map setObject:@(stat.rxAudioPacketLossSum) forKey:@"rxAudioPacketLossSum"];
  [map setObject:@(stat.rxVideoPacketLossSum) forKey:@"rxVideoPacketLossSum"];
  [map setObject:@(stat.rxAudioJitter) forKey:@"rxAudioJitter"];
  [map setObject:@(stat.rxVideoJitter) forKey:@"rxVideoJitter"];
  return map;
}

- (NSDictionary *)toLocalAudioStats:(NERtcAudioSendStats *)stat {
  NSMutableDictionary *statsMap = [[NSMutableDictionary alloc] init];
  NSMutableArray<NSDictionary *> *layers = [[NSMutableArray alloc] init];

  for (NERtcAudioLayerSendStats *layer in stat.audioLayers) {
    NSMutableDictionary *layerMap = [[NSMutableDictionary alloc] init];
    int streamType = (layer.streamType == kNERtcAudioStreamMain) ? 0 : 1;
    [layerMap setObject:@(streamType) forKey:@"streamType"];
    [layerMap setObject:@(layer.sentBitrate) forKey:@"kbps"];
    [layerMap setObject:@(layer.lossRate) forKey:@"lossRate"];
    [layerMap setObject:@(layer.rtt) forKey:@"rtt"];
    [layerMap setObject:@(layer.volume) forKey:@"volume"];
    [layerMap setObject:@(layer.numChannels) forKey:@"numChannels"];
    [layerMap setObject:@(layer.sentSampleRate) forKey:@"sentSampleRate"];
    [layerMap setObject:@(layer.capVolume) forKey:@"capVolume"];
    [layers addObject:layerMap];
  }
  [statsMap setObject:layers forKey:@"layers"];
  return statsMap;
}

- (NSDictionary *)toRemoteAudioStats:(NERtcAudioRecvStats *)stat {
  NSMutableDictionary *map = [[NSMutableDictionary alloc] init];
  [map setObject:@(stat.uid) forKey:@"uid"];

  NSMutableArray<NSDictionary *> *layers = [[NSMutableArray alloc] init];
  if (stat.audioLayers) {
    for (NERtcAudioLayerRecvStats *layer in stat.audioLayers) {
      NSMutableDictionary *layerMap = [[NSMutableDictionary alloc] init];
      int streamType = (layer.streamType == kNERtcAudioStreamMain) ? 0 : 1;
      [layerMap setObject:@(layer.receivedBitrate) forKey:@"kbps"];
      [layerMap setObject:@(layer.audioLossRate) forKey:@"lossRate"];
      [layerMap setObject:@(layer.volume) forKey:@"volume"];
      [layerMap setObject:@(layer.totalFrozenTime) forKey:@"totalFrozenTime"];
      [layerMap setObject:@(layer.frozenRate) forKey:@"frozenRate"];
      [layerMap setObject:@(streamType) forKey:@"streamType"];
      [layers addObject:layerMap];
    }
  }
  [map setObject:layers forKey:@"list"];
  return map;
}

- (NSDictionary *)toLocalVideoStats:(NERtcVideoSendStats *)stat {
  NSMutableDictionary *map = [[NSMutableDictionary alloc] init];
  NSMutableArray<NSDictionary *> *layers = [[NSMutableArray alloc] init];

  if (stat.videoLayers) {
    for (NERtcVideoLayerSendStats *layer in stat.videoLayers) {
      NSMutableDictionary *layerMap = [[NSMutableDictionary alloc] init];
      [layerMap setObject:@(layer.layerType) forKey:@"layerType"];
      [layerMap setObject:@(layer.captureWidth) forKey:@"captureWidth"];
      [layerMap setObject:@(layer.captureHeight) forKey:@"captureHeight"];
      [layerMap setObject:@(layer.width) forKey:@"width"];
      [layerMap setObject:@(layer.height) forKey:@"height"];
      [layerMap setObject:@(layer.sendBitrate) forKey:@"sendBitrate"];
      [layerMap setObject:@(layer.encoderOutputFrameRate) forKey:@"encoderOutputFrameRate"];
      [layerMap setObject:@(layer.captureFrameRate) forKey:@"captureFrameRate"];
      [layerMap setObject:@(layer.targetBitrate) forKey:@"targetBitrate"];
      [layerMap setObject:@(layer.encoderBitrate) forKey:@"encoderBitrate"];
      [layerMap setObject:@(layer.sentFrameRate) forKey:@"sentFrameRate"];
      [layerMap setObject:@(layer.renderFrameRate) forKey:@"renderFrameRate"];
      [layerMap setObject:layer.encoderName ?: @"" forKey:@"encoderName"];
      [layerMap setObject:@(layer.dropBwStrategyEnabled) forKey:@"dropBwStrategyEnabled"];
      [layers addObject:layerMap];
    }
  }
  [map setObject:layers forKey:@"layers"];
  return map;
}

- (NSDictionary *)toRemoteVideoStats:(NERtcVideoRecvStats *)stat {
  NSMutableDictionary *map = [[NSMutableDictionary alloc] init];
  [map setObject:@(stat.uid) forKey:@"uid"];

  NSMutableArray<NSDictionary *> *layers = [[NSMutableArray alloc] init];
  if (stat.videoLayers) {
    for (NERtcVideoLayerRecvStats *layer in stat.videoLayers) {
      NSMutableDictionary *layerMap = [[NSMutableDictionary alloc] init];
      [layerMap setObject:@(layer.layerType) forKey:@"layerType"];
      [layerMap setObject:@(layer.width) forKey:@"width"];
      [layerMap setObject:@(layer.height) forKey:@"height"];
      [layerMap setObject:@(layer.receivedBitrate) forKey:@"receivedBitrate"];
      [layerMap setObject:@(layer.fps) forKey:@"fps"];
      [layerMap setObject:@(layer.packetLossRate) forKey:@"packetLossRate"];
      [layerMap setObject:@(layer.decoderOutputFrameRate) forKey:@"decoderOutputFrameRate"];
      [layerMap setObject:@(layer.rendererOutputFrameRate) forKey:@"rendererOutputFrameRate"];
      [layerMap setObject:@(layer.totalFrozenTime) forKey:@"totalFrozenTime"];
      [layerMap setObject:@(layer.frozenRate) forKey:@"frozenRate"];
      [layerMap setObject:layer.decoderName ?: @"" forKey:@"decoderName"];
      [layers addObject:layerMap];
    }
  }
  [map setObject:layers forKey:@"layers"];
  return map;
}

- (NSDictionary *)toNetworkQualityStats:(NERtcNetworkQualityStats *)stat {
  NSMutableDictionary *map = [[NSMutableDictionary alloc] init];
  [map setObject:@(stat.userId) forKey:@"uid"];
  [map setObject:@(stat.txQuality) forKey:@"txQuality"];
  [map setObject:@(stat.rxQuality) forKey:@"rxQuality"];
  return map;
}

@end